access/validator.go

```diff
@@ -346,6 +386,54 @@ func (v *TransactionValidator) checkSignatureFormat(tx *flow.TransactionBody) er
 	return nil
 }
 
+func (v *TransactionValidator) checkSufficientBalanceToPayForTransaction(ctx context.Context, tx *flow.TransactionBody) error {
+	if !v.options.CheckPayerBalance {
+		return nil
+	}
+
+	header, err := v.blocks.SealedHeader()
+	if err != nil {
+		return fmt.Errorf("could not fetch block header: %w", err)
+	}
+
+	payerAddress := cadence.NewAddress(tx.Payer)
+	inclusionEffort := cadence.UFix64(tx.InclusionEffort())
+	gasLimit := cadence.UFix64(tx.GasLimit)
+
+	args, err := cadenceutils.EncodeArgs([]cadence.Value{payerAddress, inclusionEffort, gasLimit})
+	if err != nil {
+		return fmt.Errorf("failed to encode cadence args for script executor: %w", err)
+	}
+
+	result, err := v.scriptExecutor.ExecuteAtBlockHeight(ctx, v.verifyPayerBalanceScript, args, header.Height)
+	if err != nil {
+		return fmt.Errorf("script finished with error: %w", err)
+	}
+
+	value, err := jsoncdc.Decode(nil, result)
+	if err != nil {
+		return fmt.Errorf("could not decode result value returned by script executor: %w", err)
+	}
+
+	canExecuteTransaction, requiredBalance, _, err := fvm.DecodeVerifyPayerBalanceResult(value)
+	if err != nil {
+		return fmt.Errorf("could not parse cadence value returned by script executor: %w", err)
+	}
+
+	// return no error if payer has sufficient balance
+	if bool(canExecuteTransaction) {
+		return nil
+	}
+
+	account, err := v.scriptExecutor.GetAccountAtBlockHeight(ctx, tx.Payer, header.Height)
```

Comment:
sorry for the confusion. I was thinking the 3rd response value from `DecodeVerifyPayerBalanceResult` was the account's current balance. If we have to lookup the balance explicitly, I don't think we should do it.